src/acl.c

```diff
@@ -1238,179 +1491,358 @@ user *ACLGetUserByName(const char *name, size_t namelen) {
+    if (flags & CMD_KEY_DELETE) key_flags |= ACL_WRITE_PERMISSION;
+    if (flags & CMD_KEY_UPDATE) key_flags |= ACL_WRITE_PERMISSION;
 
     /* Test this key against every pattern. */
+    while((ln = listNext(&li))) {
+        keyPattern *pattern = listNodeValue(ln);
+        if ((pattern->flags & key_flags) != key_flags)
+            continue;
+        size_t plen = sdslen(pattern->pattern);
+        if (stringmatchlen(pattern->pattern,plen,key,keylen,0)) 
+            return ACL_OK;
+    }
+    return ACL_DENIED_KEY;
+}
+
+/* Returns if a given command may possibly access channels. For this context,
+ * the unsubscribe commands do not have channels. */
+static int ACLDoesCommandHaveChannels(struct redisCommand *cmd) {
+    return (cmd->proc == publishCommand
+        || cmd->proc == subscribeCommand
+        || cmd->proc == psubscribeCommand
+        || cmd->proc == spublishCommand
+        || cmd->proc == ssubscribeCommand);
+}
+
+/* Checks a channel against a provide list of channels. */
+static int ACLCheckChannelAgainstList(list *reference, const char *channel, int channellen, int literal) {
+    listIter li;
+    listNode *ln;
+
+    listRewind(reference, &li);
     while((ln = listNext(&li))) {
         sds pattern = listNodeValue(ln);
         size_t plen = sdslen(pattern);
-        if (stringmatchlen(pattern,plen,key,keylen,0))
+        if ((literal && !strcmp(pattern,channel)) || 
+            (!literal && stringmatchlen(pattern,plen,channel,channellen,0)))
+        {
             return ACL_OK;
+        }
     }
-    return ACL_DENIED_KEY;
+    return ACL_DENIED_CHANNEL;
 }
 
-/* Check if the command is ready to be executed according to the
- * ACLs associated with the specified user.
+/* Check if the pub/sub channels of the command can be executed
+ * according to the ACL channels associated with the specified selector.
+ * 
+ * idx and count are the index and count of channel arguments from the
+ * command. The literal argument controls whether the selector's ACL channels are 
+ * evaluated as literal values or matched as glob-like patterns.
  *
- * If the user can execute the command ACL_OK is returned, otherwise
- * ACL_DENIED_CMD or ACL_DENIED_KEY is returned: the first in case the
- * command cannot be executed because the user is not allowed to run such
- * command, the second if the command is denied because the user is trying
- * to access keys that are not among the specified patterns. */
-int ACLCheckCommandPerm(const user *u, struct redisCommand *cmd, robj **argv, int argc, int *keyidxptr) {
-    int ret;
-    uint64_t id = cmd->id;
+ * If the selector can execute the command ACL_OK is returned, otherwise 
+ * ACL_DENIED_CHANNEL. */
+static int ACLSelectorCheckPubsubArguments(aclSelector *s, robj **argv, int idx, int count, int literal, int *idxptr) {
+    for (int j = idx; j < idx+count; j++) {
+        if (ACLCheckChannelAgainstList(s->channels, argv[j]->ptr, sdslen(argv[j]->ptr), literal != ACL_OK)) {
```

Comment:
I don't split, and I've seen others not explicitly try to split it as well, when it doesn't really help readability to split.

Reply:
i think it damages readability.
- splitting at arbitrary position (offset, rather than a logical one).
- adds more lines so we get less lines fit in our screen.
- these days people have very wide monitors (and vertical screen space is needed, and there's an abundance of unused horizontal space)